proposals/split_kyverno_controllers.md

```diff
@@ -0,0 +1,120 @@
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Definitions](#definitions)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+- [CRD Changes (OPTIONAL)](#crd-changes-optional)
+
+# Overview
+[overview]: #overview
+
+Kyverno 1.8 runs as a single Deployment. This includes the webhook, reporting, and background controllers for generate and update requests. This design causes heavy backhround processing to impact the webhook performance. This proposal separates the webhook into its own Deployment.
+
+# Definitions
+[definitions]: #definitions
+
+Make a list of the definitions that may be useful for those reviewing. Include phrases and words that Kyverno users or other interested parties may not be familiar with.
+
+# Motivation
+[motivation]: #motivation
+
+- Eliminate, or greatly reduce, client-side throttling of the webhook.
+- Improve webhook performance
+- Allow independent sizing and scalability of webhook and background processing controllers
+
+# Proposal
+
+## Kyverno Controllers
+
+### Webhook Controller
+
+The webhook controller is responsible for:
+1. Processing AdmissionReview requests from the API Server
+2. Listening to policy and policyexception changes to maintain the policy cache
+
+### Background Controller
+
+The background controller is responsible for:
+1. Processing report changes
+2. Processing mutate and generate requests
+
+**NOTE 1**: its possible to further decompose the background controller, but seems like that may be of incremental value. Even with a single background controller, we can enable / disable modules or features.
+
+**NOTE 2**: we need to decide if the new cleanup controller is part of the background controller or should be separated.
+
+## Code Repository Structure
+
+All Kyverno code can stay in the same repository, and the top-level packages can reflect the logical design:
+
+* **controllers/webhook**
+* **controllers/background**
+
+## User Experience
+
+This section describes the user experience across installation and management of Kyverno:
+
+### Installation
+
+Kyverno can still be deployed as a single Helm chart or via a single YAML file. 
+
+### Runtime 
+
+When Kyverno is installed there would be two Deployment resources created. The Helm chart can allow each Deployment to be scaled independently. 
+
+### Configuration
+
+The Kyverno ConfigMap can allow enabling or disabling all major features or modules.
+
+# Implementation
+
+TBD.
+
+## Link to the Implementation PR
+
+# Migration (OPTIONAL)
+
+When a new version of Kyverno, with this feature, is installed it will deploy two controllers. The `kyverno` deployment can continue to be the webhook, and a new `kyverno-background-controller` can execute to process the reports and update requests. 
```

Comment:
This could lead to a breaking change, some flags will be removed from kyverno.
We also need to pay attention to leader election, some controllers currently running in kyverno are only running in the leader pod, if we start running the same controllers in separate binary there's a chance that two controllers of the same type will fight against each other.

Reply:
> This could lead to a breaking change, some flags will be removed from kyverno.

Can you provide an example?

> This could lead to a breaking change, some flags will be removed from kyverno.
We also need to pay attention to leader election, some controllers currently running in kyverno are only running in the leader pod, if we start running the same controllers in separate binary there's a chance that two controllers of the same type will fight against each other.

But in a rolling upgrade, wouldn't the new instance take over via a leader election? Seems like its the same when you have binaries of different versions. I am not sure why it matters if the instances are in the same binary or different ones, as the leader election is coordinated via the lease. We would need to make sure the same lease is used across the migration.